### PR TITLE
Add LIIF.

### PR DESCRIPTION
**Pull request by Yshuo-Li**
Add LIIF model for single image super-resolution.
Paper: Learning Continuous Image Representation with Local Implicit Image Function
Ref repo: https://github.com/yinboc/liif.git

### CODE REVIEW

**Comment by innerlee** [comment]
I will take a look tonight :)

File: mmedit/models/restorers/liif.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                 feat_unfold=True,
+                 cell_decode=True,
+                 data_mean=[0.5],
+                 data_std=[0.5],
+                 eval_bsize=None,
+                 pixel_loss=None,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        super(BasicRestorer, self).__init__()
```
Comment (innerlee) [inline mmedit/models/restorers/liif.py:58]: Double check L58

**Comment by innerlee** [comment]
There are some thing wrong with commits.

File: mmedit/datasets/pipelines/utils.py [hunk 2]
```diff
@@ -129,6 +129,14 @@ def random_choose_unknown(unknown, crop_size):
...
 def make_coord(shape, ranges=None, flatten=True):
     """ Make coordinates at grid centers.
+
+    Args:
+        shape (tuple): shape of image.
+        ranges (tuple): range of coordinate value. Default: None.
+        flatten (bool): flatten to (n, 2) or Not. Default: True.
+
+    return:
+        coord (Tensor): coordinates.
```
Comment (innerlee) [inline mmedit/datasets/pipelines/utils.py:139]: What's the shape of coord

Comment (Yshuo-Li) [reply mmedit/datasets/pipelines/utils.py:139]: Assume shape = h, w
if flatten: coord.shape = [h, w, 2]
else: coord.shape = [h*w, 2]

File: mmedit/models/restorers/liif.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        imnet['in_dim'] = imnet_in_dim
+        self.imnet = build_component(imnet)
+
+        # support fp16
+        self.fp16_enabled = False
+
+        # loss
+        self.pixel_loss = build_loss(pixel_loss) if pixel_loss else None
+
+        self.step_counter = 0  # counting training steps
```
Comment (innerlee) [inline mmedit/models/restorers/liif.py:98]: What if the model is resumed

File: mmedit/models/restorers/liif.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        Args:
+            lq (Tensor): Input lq images.
+            gt (Tensor): Ground-truth image. Default: None.
+            test_mode (bool): Whether in test mode or not. Default: False.
+            kwargs (dict): Other arguments.
+
+        Returns:
+            result (Tensor): (part of) output.
+        """
+        # result = {**kwargs}
```
Comment (innerlee) [inline mmedit/models/restorers/liif.py:383]: remove

File: mmedit/models/restorers/liif.py (hunk 3 above)
Comment (Yshuo-Li) [reply mmedit/models/restorers/liif.py:98]: This parameter (`step_counter`) is actually not used and has been deleted.

File: mmedit/models/restorers/liif.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        """Validation step.
+
+        Args:
+            data_batch (dict): A batch of data.
+            kwargs (dict): Other arguments for ``val_step``.
+
+        Returns:
+            dict: Returned output.
+        """
+        # print(data_batch.keys())
```
Comment (innerlee) [inline mmedit/models/restorers/liif.py:243]: remove debuggin codes

**Review by innerlee — APPROVED** [review]

**Pull request merged**